src/Drivers/HubbardDriver.cpp

```diff
@@ -0,0 +1,59 @@
+#include "Drivers/HubbardDriver.hpp"
+
+namespace GQCP {
+
+    HubbardDriver::HubbardDriver(std::string csline, size_t num_states, size_t num_alpha, size_t num_beta, size_t K) {
+
+        std::string triagonal_line = csline;
+
+        // Read the upper triagonal of the hopping matrix
+        GQCP::VectorX<double> triagonal;
+        if (csline.empty()) {
+            throw std::invalid_argument("hubbard(driver): Comma-separated was empty!");
+        }
+
+        // Split comma-separated line
+        std::vector<std::string> splitted_line;
+        boost::split(splitted_line, triagonal_line, boost::is_any_of(","));
+
+        std::vector<double> triagonal_data;
+        for (const std::string &x : splitted_line) {
+            triagonal_data.push_back(std::stod(x));
+        }
+
+        triagonal = Eigen::Map<Eigen::VectorXd>(triagonal_data.data(), triagonal_data.size());
+        GQCP::HoppingMatrix H = GQCP::HoppingMatrix::FromUpperTriangle(triagonal);
+
+        // Actual calculations
+        auto ham_par = GQCP::HamiltonianParameters<double>::Hubbard(H);
+        if (ham_par.get_K() != K) {
+            throw std::invalid_argument("hubbard(driver): The given number of sites does not match the triagonal");
+        }
+
+
+        // Initialize and solve the Hubbard eigenvalue problem
```

Comment:
Since only a `solve` method would use the `ProductFockSpace`, `Hubbard` Hamiltonian builder and `CISolver`, I don’t think it is necessary to include these as members of the class (they don’t need to exist outside of the actual calculation part).

Removing these objects as members from `HubbardDriver` would bypass the issue that you tried to solve by using `shared_ptr`s.